### PR TITLE
Add a signal to notify when children nodes enter or exit tree

### LINKED ISSUE #724: Add `node_added` and `node_removed` signals to Node (in addition to the existing SceneTree signals)
**Describe the project you are working on:**
2D Shooters.

**Describe the problem or limitation you are having in your project:**
Tool code API  and behavior issues.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It would be useful in many tool scripts to be alerted  when a Node's descending Nodes have been added or removed using a signal.

Currently, these signals only exists at the level of the **SceneTree**.

Within the editor this signal fires quite frequently, as the editor itself is part of the tree.  This leads any filtering logic to have to constantly filter through hundreds of nodes, and can create serious performance issues.

It makes it difficult, error prone, and costly.

Would be better if each node could report what nodes have been added or removed from their hierarchy.

There are certainly use cases for this in run-time scripts as well, but I'll have to leave that to others to provide, as my development preference is to try to avoid adding/removing nodes from the scene tree as much as possible.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

The signals might be on Node, and the GDScript would follow as:

```
tool
extends Node2D

func _ready():

	connect("node_added", self, "node_added")
	connect("node_removed", self, "node_removed")
```

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

It usually ends up being a dozen or some lines with the most minimal work around.  The issue though is mainly one of performance.  Having to examine and act on all these editor related nodes with each interaction in the editor.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Two built-in signals would be far more convenient than rebuilding with a module, and the plugin wouldn't solve the performance issue.

### LINKED ISSUE #2544: Add a `Node.node_added` signal/method to detect the addition of children and subchildren nodes
### Describe the project you are working on
A project based of components added to a node

### Describe the problem or limitation you are having in your project
In my project, a custom node can have multiple components that describes it's behaviour, each component can be added as the direct child or subchild of this node. I need a way to easily tell this node that a child or subchild node was added, even if not inside tree yet, something that `SceneTree.node_added` cannot do.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The easiest way I could think to tell the node that a direct/indirect child node was added, was to use some signal or virtual method. Something like `Node.node_added(node: Node, parent: Node)`. This could be the same to when a direct/indirect child node was removed, something like `Node.node_removed(node: Node, prevParent: Node)`, the second parameter of the signal/method would be the previous parent of the removed node. It's more interesting to be a virtual method, so it could be called even when is not added to the tree yet.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Let's say that I have this tree:
image
And in some point, a node called Node6 is added to Node5, and I want to tell the Node1 that this new node was added, then from SceneTree would call a notification, signal or method to tell the parents and grandparents that this node was added. The same would be if this node was removed from the tree.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I'm not sure if there's a better way to do this, but it would be a nice addition to the engine IMO.

### Is there a reason why this should be core and not an add-on in the asset library?
Because it requires to modify the core structure of a Node, to be able to add this behaviour.

### PR DESCRIPTION
**Pull request by reduz**
* Allows more fine grained notifications (hence better performance) than using the global scene tree signals (node added and removed).
* Required for #55950

*Bugsquad edit:*
- Resolves https://github.com/godotengine/godot-proposals/issues/724
- Resolves https://github.com/godotengine/godot-proposals/issues/2544

### CODE REVIEW

**Review by mhilbrunner — APPROVED** [review]

**Comment by sairam4123** [comment]
Could this be backported to Godot 3.x?

**Comment by hilfazer** [comment]
What about a notification for this (perhaps as an addition to a signal)? There is one for a node getting parented.

**Comment by akien-mga** [comment]
We discussed this at length on Rocket.chat and in a PR review meeting, so I'll summarize some of the points we discussed to validate this core change:

- The use case this addresses is for an external node (an observer) to monitor changes in the children nodes of a given node. This should be done without having to add code to the monitored node or its children, so all `_notification` messages are not sufficient, as those can only be handled by the node which receives the notification (the child node, or its parent - not the observer).
- This use case _can_ be addressed using the `SceneTree.node_added` signal, but this signal is emitted for every single node added to the scene tree - it has a huge performance cost and is only really relevant for debugging / editor tools (that performance cost is only paid once a callback is connected to the signal - but that's what would be needed here).
- We raised concern that we have *a lot* of ways to tap into the enter tree / ready / exit tree states. They all fulfill different needs but it would still be worth trying to make some visualization of these APIs and see if there isn't some redundancy that can be reduced. From the top of my head:
  * `SceneTree`:
    - Signals: `node_added`, `node_removed`, `node_renamed`, `tree_changed` (expensive, mostly for debugging)
  * `Node`:
    - Own state signals: `tree_entered`, `ready`, `tree_exiting`, `tree_exited`, `renamed` (can be connected to from outside to monitor the state of a given node)
    - Child state signals: `child_entered_tree` and `child_exited_tree` added in this PR (can be connected to from outside to monitor the state of a given node's direct children)
    - Notifications: `NOTIFICATION_ENTER_TREE`, `NOTIFICATION_READY`, `NOTIFICATION_EXIT_TREE`, `NOTIFICATION_MOVED_IN_PARENT`, `NOTIFICATION_PARENTED`, `NOTIFICATION_UNPARENTED` (these are all received by the node itself when it hits each of these conditions - external monitors don't have access to this)
    - Callbacks: `_enter_tree`, `_ready`, `_exit_tree` (these are convenience virtual methods which are equivalent to handling `NOTIFICATION_ENTER_TREE`, `NOTIFICATION_READY` and `NOTIFICATION_EXIT_TREE` in `_notification(what)`.
- We discussed that the expensive `SceneTree.node_added` could maybe be moved to `Node` so that it can be used to monitor a tree branch (a node, its children, and all their children recursively). This was considered not to be a good option for the specific use case defined above (where only direct children are relevant), as it would still be quite expensive. It's only not clear how easy to implement it would be (`SceneTree.node_added` is just each Node telling the common `SceneTree` to propagate the signal - if moved to Node then all parent nodes need to be notified? which is also quite expensive).
- For users who want to keep track of all children and grandchildren recursively in a branch, the signals added in this method can still be used, you will just have to connect them recursively (i.e. for each direct child, connect to their `child_entered_tree` signals, etc.).

TL;DR: A lot of discussion because this is a pretty core API change and we want to do this right. But all things considered that seems to be a pretty good solution for a use case which was not yet well covered by the existing, somewhat redundant APIs.

**Review by akien-mga — APPROVED** [review]
Approved in PR review meeting, see above summary of the discussion.

**Comment by akien-mga** [comment]
> Could this be backported to Godot 3.x?

I think so yes.

> What about a notification for this (perhaps as an addition to a signal)? There is one for a node getting parented.

I answer this in part in my big paragraph above. A notification like `NOTIFICATION_CHILD_ADDED` would only be received by the parent node, and therefore it would still not be suitable for this use case where an external node needs to monitor the parent node that got a new child (i.e. without changing the script of the parent node).

That _could_ work if we added a way for external nodes to subscribe to notifications for specific nodes, which could be an avenue to explore to reduce the redundancy between existing notifications and their matching signals... but I'm not sure how feasible and meaningful that would be, it could be a topic for a dedicated proposal/discussion.

**Comment by Faless** [comment]
> * Notifications: `NOTIFICATION_ENTER_TREE`, `NOTIFICATION_READY`, `NOTIFICATION_EXIT_TREE`, `NOTIFICATION_MOVED_IN_PARENT`, `NOTIFICATION_PARENTED`, `NOTIFICATION_UNPARENTED` (these are all received by the node itself when it hits each of these conditions - external monitors don't have access to this)

There is also `NOTIFICATION_POST_ENTER_TREE`, which is fired right before `NOTIFICATION_READY` but fires every time the child enters tree, unlike `NOTIFICATION_READY` which is only called once per node.
That too, is not exposed externally, but would be useful in the replication PR (and in cases when you want to track nodes).
But I agree that we have so many signals already. maybe it's worth exploring the idea of exposing a `node_notification` signal that groups them all (passing the `NOTIFICATION_*` parameter).

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.

**Comment by samdze** [comment]
Does this PR make signals to be emitted when a child is moved from the parent to another node?
Doesn't seem like so.

**Comment by Zireael07** [comment]
@samdze: "Moving" from parent to another node (i.e. reparenting) never fires enter/exit callbacks, that's add or remove child

**Comment by akien-mga** [comment]
@samdze Works fine here:

```gdscript
# main.gd
extends Node

func _ready():
	var child = Node.new()
	child.name = "Child"
	print("Adding Child to Node1")
	$Node1.add_child(child)
	print("Removing Child from Node1")
	$Node1.remove_child(child)
	print("Adding Child to Node2")
	$Node2.add_child(child)
	print("Removing Child from Node2")
	$Node2.remove_child(child)
```

```gdscript
# node.gd (used by Node1 and Node2)
extends Node

func on_child_entered_tree(node: Node):
	print("Child node '%s' entered tree in parent '%s'." % [node.name, name])

func on_child_exiting_tree(node: Node):
	print("Child node '%s' exiting tree in parent '%s'." % [node.name, name])

func _ready():
	child_entered_tree.connect(on_child_entered_tree)
	child_exiting_tree.connect(on_child_exiting_tree)
```

Prints:
```
Adding Child to Node1
Child node 'Child' entered tree in parent 'Node1'.
Removing Child from Node1
Child node 'Child' exiting tree in parent 'Node1'.
Adding Child to Node2
Child node 'Child' entered tree in parent 'Node2'.
Removing Child from Node2
Child node 'Child' exiting tree in parent 'Node2'.
```

[TestChildNotify.zip](https://github.com/godotengine/godot/files/9044026/TestChildNotify.zip)

**Comment by samdze** [comment]
Yes that's what I thought, reparenting a node doesn't trigger these signals.
That's too bad, nodes that depend on this functionality to manage children may end up with nodes reparented to themselves, without a signal triggered, or reparented outside, again without a signal triggered. 
But I understand this should be a rather rare use case.

Removing and re-adding a node it's another option but that's not something the user does spontaneously.

**Comment by akien-mga** [comment]
> Yes that's what I thought, reparenting a node doesn't trigger these signals.

What do you mean by reparenting a node?
To reparent a node, you have to remove it from its parent, and then add it to its new parent. That's what my example does. There's no `reparent` method but if there was it would do exactly that.

**Comment by samdze** [comment]
Whoops!
You're right!
